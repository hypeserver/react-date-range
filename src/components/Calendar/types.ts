import React from "react";


export interface Scroll {
  enabled: boolean
  monthWidth?: number | string
  monthHeight?: number | string
  longMonthHeight?: number | string
  // FIXME
  calendarWidth?: any
  calendarHeight?: any
}

export enum CalendarDirection {
  VERTICAL = 'vertical',
  HORIZONTAL = 'horizontal'
}


export enum ChangeDateMode {
  SET = 'set',
  MONTH_OFFSET = 'monthOffset',
  SET_MONTH = 'setMonth',
  SET_YEAR = 'setYear'
}


export interface AriaLabelsShape {
  dateInput: { startDate: string, endDate: string }[]
  monthPicker: string,
  yearPicker: string,
  prevButton: string,
  nextButton: string,
}
