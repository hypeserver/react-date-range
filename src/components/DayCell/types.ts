import { DateRange } from "../../utilsTypes";

export interface ExtendedDateRange extends DateRange {
  isStartEdge: Boolean
  isEndEdge: Boolean
  isInRange: Boolean
}

export interface Preview extends DateRange {
  range?: DateRange
}

export interface Styles {
  calendarWrapper?: string
  dateDisplay?: string
  dateDisplayItem?: string
  dateDisplayWrapper?: string
  day?: string
  dayActive: string
  dayDisabled: string
  dayEndOfMonth: string
  dayEndOfWeek: string
  dayEndPreview: string
  dayHovered: string
  dayInPreview: string
  dayNumber?: string
  dayPassive: string
  days?: string
  dayStartOfMonth: string
  dayStartOfWeek: string
  dayStartPreview: string
  dateRangePickerWrapper?: string
  dayToday: string
  dayWeekend: string
  endEdge: string
  infiniteMonths?: string
  inRange: string
  month?: string
  monthAndYearDivider?: string
  monthAndYearPickers?: string
  monthAndYearWrapper?: string
  monthName?: string
  monthPicker?: string
  months?: string
  monthsHorizontal?: string
  monthsVertical?: string
  nextButton?: string
  nextPrevButton?: string
  prevButton?: string
  selected?: string
  startEdge: string
  weekDay?: string
  weekDays?: string
  yearPicker?: string
  dateRangeWrapper?: string
  inputRange?: string
  inputRangeInput?: string
  inputRangeLabel?: string
  // FIXME
  dateDisplayItemActive?: any
}
