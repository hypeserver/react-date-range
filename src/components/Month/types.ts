import { DateRange } from "../../utilsTypes";



export interface Drag {
  range: DateRange
  disablePreview: boolean
  status: boolean
}
