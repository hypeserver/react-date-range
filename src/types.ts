// Type definitions for react-date-range-won 1.5
// Project: https://github.com/gbili/react-date-range-won (Does not have to be to GitHub, but prefer linking to a source code repository rather than to a project website.)
// Definitions by: Guillermo Pages <https://github.com/gbili>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

import React from 'react';
import { Locale } from 'date-fns';
import { ClassNames } from './styles';

export type AnyDate = string | Date;
export type DateOrTimestamp = Date | number;
export type DateFunc = (now: Date) => AnyDate;
export type DateGen = (now: Date) => Date;
export type DateInputType = AnyDate | DateGen;
export type LanguageType = 'cn' | 'jp' | 'fr' | 'it' | 'de' | 'ko' | 'es' | 'ru' | 'tr';
export type SizeType = number;

export interface DateContainerType {
  date: Date;
}

export type AriaLabelShape = {
  dateInput?: {
    [x: string]: StartEndDate<Date | string>;
  };
  monthPicker?: string;
  yearPicker?: string;
  prevButton?: string;
  nextButton?: string;
}


export interface CalendarTheme {
  DateRange?: React.CSSProperties;
  Calendar?: React.CSSProperties;
  Day?: React.CSSProperties;
  DayPassive?: React.CSSProperties;
  DayHover?: React.CSSProperties;
  DayToday?: React.CSSProperties;
  DaySunday?: React.CSSProperties;
  DaySpecialDay?: React.CSSProperties;
  DayActive?: React.CSSProperties;
  DaySelected?: React.CSSProperties;
  DayStartEdge?: React.CSSProperties;
  DayEndEdge?: React.CSSProperties;
  DayInRange?: React.CSSProperties;
  Weekday?: React.CSSProperties;
  MonthAndYear?: React.CSSProperties;
  MonthButton?: React.CSSProperties;
  MonthArrow?: React.CSSProperties;
  MonthArrowPrev?: React.CSSProperties;
  MonthArrowNext?: React.CSSProperties;
  PredefinedRanges?: React.CSSProperties;
  PredefinedRangesItem?: React.CSSProperties;
  PredefinedRangesItemActive?: React.CSSProperties;
}

export type ModeMapper = {
  monthOffset: () => Date,
  setMonth: () => Date,
  setYear: () => Date,
  set: () => number,
}

export function isModeMapperKey(s: string, o: ModeMapper): s is keyof ModeMapper {
  return Object.keys(o).indexOf(s) !== -1;
}

export interface RangeWithKey extends Range {
  key: 'selection';
}

export interface OnDateRangeChangeProps {
  [key: string]: Range;
}

export type WeekStartsOn = 0|1|2|3|4|5|6;

export type DateOptions = { locale: Locale; weekStartsOn?: WeekStartsOn; };

export function isRangeValue(value: Range | Date): value is Range {
  return value.hasOwnProperty('startDate') && value.hasOwnProperty('endDate');
}

export function isOnlyWithStartDate(value: Range | Date): value is Range {
  return value.hasOwnProperty('startDate') && value.hasOwnProperty('endDate');
}
export function isSureRange<T>(range: Range & T): range is SureStartEndDate & T {
  return range.startDate instanceof Date && range.endDate instanceof Date;
}

export interface CommonCalendarProps {
  /** default: DD/MM/YYY */
  format?: string;
  firstDayOfWeek?: number;
  theme?: CalendarTheme;
  /** default: none */
  onInit?: ((range: Range) => void);
  /** default: none */
  minDate?: Date;
  /** default: none */
  maxDate?: Date;
  /** default: */
  weekStartsOn?: WeekStartsOn;
  /** default: enUs from locale. Complete list here https://github.com/Adphorus/react-date-range/blob/next/src/locale/index.js */
  locale?: Locale;
  /** Custom class names for elements */
  classNames?: Partial<ClassNames>;
  /** default: none */
  navigatorRenderer?: (
        currentFocusedDate: Date,
        changeShownDate: (shownDate: Date) => void,
        props: CommonCalendarProps
      ) => JSX.Element;
  /** default: none */
  onShownDateChange?: ((visibleMonth: Date) => void);
  /** default: none */
  onRangeFocusChange?: ((focusedRange: RangeFocus) => void);
  /** default: false */
  editableDateInputs?: boolean;
  /** default: true */
  dragSelectionEnabled?: boolean;
  /** default: false */
  fixedHeight?: boolean;
}

export type DisplayMode = 'date' | 'dateRange';
export type CalendarFocus = 'forwards' | 'backwards';
export type CalendarDirection = 'vertical' | 'horizontal';

export interface CalendarProps extends CommonCalendarProps {
  /** default: today */
  date: DateInputType;
  /** default: none */
  onChange?: (date?: Date) => void;
  scroll?: { enabled: boolean; };
  preventSnapRefocus?: boolean;
  calendarFocus: CalendarFocus;
  months: number;
}

export class Calendar extends React.Component<CalendarProps> {}

export interface DateRangeProps extends Range, CommonCalendarProps {
  /** default: enUs from locale. Complete list here https://github.com/Adphorus/react-date-range/blob/next/src/locale/index.js */
  locale?: Locale;
  /** default: false */
  linkedCalendars?: boolean;
  /** default: 2 */
  calendars?: number;
  /** default: none */
  ranges?: Range[];
  /** default: { enabled: false } */
  scroll?: ScrollOptions;
  /** default: false */
  showSelectionPreview?: boolean;
  /** default: false */
  twoStepChange?: boolean;
  /** default: true */
  showMonthArrow?: boolean;
  /** default: false */
  rangedCalendars?: boolean;
  /** default: none */
  specialDays?: DateContainerType[];
  /** default: 1 */
  months?: number;
  /** default: true */
  showMonthAndYearPickers?: boolean;
  /** default: [] */
  rangeColors?: string[];
  /** default: */
  shownDate?: Date;
  /** default: */
  disabledDates?: Date[];
  /** default: */
  disabledDay?: ((date: Date) => boolean);
  /** default: Early */
  startDatePlaceholder?: string;
  /** default: */
  className?: string;
  /** default: Continuous */
  endDatePlaceholder?: string;
  /** default: MMM d, yyyy */
  dateDisplayFormat?: string;
  /** default: d */
  dayDisplayFormat?: string;
  /** default: E */
  weekdayDisplayFormat?: string;
  /** default: MMM yyyy */
  monthDisplayFormat?: string;
  /** default: vertical */
  direction?: CalendarDirection;
  /** default: false */
  moveRangeOnFirstSelection?: boolean;
  /** default: false */
  retainEndDateOnFirstSelection?: boolean;
  /** default: false */
  editableDateInputs?: boolean;
  /** default: */
  focusedRange?: RangeFocus;
  /** default: [0, 0] */
  initialFocusedRange?: RangeFocus;
  /** default: */
  onRangeFocusChange?: ((focusedRange: RangeFocus) => void);
  /** default: */
  preview?: Preview;
  /** default: true */
  showPreview?: boolean;
  /** default: */
  // onPreviewChange?: (preview?: Preview) => void;
  /** default: none */
  onChange?: (range: OnDateRangeChangeProps) => void;
}

export interface DateRangePickerProps extends DateRangeProps {
  renderStaticRangeLabel?: ((range: DefinedRange) => JSX.Element);
  staticRanges?: StaticRange[];
  inputRanges?: InputRange[];
}

export class DateRange extends React.Component<DateRangeProps> {}

export class DateRangePicker extends React.Component<DateRangePickerProps> {}

export type DateRangeIndex = 'Today' | 'Yesterday' | 'Last 7 Days' | 'Last 30 Days';

export type SureStartEndDate<D = Date> = {
  startDate: D;
  endDate: D;
}

export type SureStartMaybeEndDate<D = Date> = {
  startDate: D;
  endDate: D | null;
}

export type SureStartNoEndDate<D = Date> = {
  startDate: D;
  endDate: null;
}

export type MaybeMaybeRange<D = Date> = OtherRangeProps & (SureStartEndDate<D> | SureStartMaybeEndDate<D> | SureStartNoEndDate<D>);

export function isNoEndDateRange(r: MaybeMaybeRange): r is SureStartNoEndDate {
  return r.hasOwnProperty('startDate') && r.endDate === null;
}

export type StartEndDate<D = Date> = Partial<SureStartEndDate<D>>;

export type StartEndDateGen = () => SureStartEndDate;
export type RangeGen = (props?: CommonCalendarProps) => Range;

export type WithSureRangeGen = { range: StartEndDateGen; };
export type WithRangeGen = { range: RangeGen; };
export type WithRange = { range: Range; };
export type WithRangeOrRangeGen = { range: Range | RangeGen; };

export function isWithRangeGen(wr: WithRangeOrRangeGen): wr is WithRangeGen {
  return typeof wr.range === 'function';
}
export function isWithRange(wr: WithRangeOrRangeGen): wr is WithRange {
  return typeof wr.range !== 'function';
}

export type LabeledStartEndDateGen = WithSureRangeGen & {
  label: string;
}

export type Json =
  | null
  | string
  | number
  | boolean
  | Array<JSON>
  | {
    [prop: string]: Json
  }

export interface OtherRangeProps {
  color?: string;
  key?: string;
  autoFocus?: boolean;
  disabled?: boolean;
  showDateDisplay?: boolean;
}

export interface SureRange<D = Date> extends SureStartEndDate<D>, OtherRangeProps {}

export interface Range<D = Date | null> extends StartEndDate<D>, OtherRangeProps {}

export interface ScrollOptions {
  enabled: boolean;
  monthHeight?: number;
  longMonthHeight?: number;
  monthWidth?: number;
  calendarWidth?: number;
  calendarHeight?: number;
}

export type WithLabel = { label: string; };
export type WithMaybeLabel = { label?: string; };
export type WithMaybeId = { id?: string; };
export type WithIsSelected = { isSelected: (range: SureStartEndDate) => boolean; }
export interface DefinedRangeCommon extends WithMaybeLabel, WithMaybeId {
  hasCustomRendering?: boolean;
}

export type WithRangeCallback = { range: (props?: CommonCalendarProps) => SureRange; }

export interface StaticRange extends DefinedRangeCommon, WithRangeOrRangeGen, WithIsSelected {
}

export type StaticRangeWihLabel = StaticRange & WithLabel & WithIsSelected;

export interface InputRange {
  range: (value: Date | number, props?: CommonCalendarProps) => Range;
  getCurrentValue: (range: Range) => number | "-" | "âˆž";
}
export type InputRangeWihLabel = InputRange & WithLabel;

export type DefinedRange = StaticRange | InputRange;

/**
 * Represents range focus `[range, rangeElement]`. `range` represents the index of the range
 * that's focused and the `rangeElement` the element of the range that's
 * focused, `0` for start date and `1` for end date
 */
export type RangeFocus = [number, number];

export type Preview = MaybeMaybeRange & {
  color?: string;
}
