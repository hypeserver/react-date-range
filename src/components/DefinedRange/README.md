#### Example: Default Labels

```jsx inside Markdown
import { useState } from 'react';

const [state, setState] = useState([
  {
    startDate: new Date(),
    endDate: null,
    key: 'selection'
  }
]);

<DefinedRange
  onChange={item => setState([item.selection])}
  ranges={state}
/>;
```
#### Example: Custom range labels

```jsx inside Markdown
import { useState, useEffect } from 'react';

const CustomStaticRangeLabelContent = ({ text }) => {
  const [currentDateString, setCurrentDateString] = useState(Date());

  useEffect(() => {
    const intervalId = setInterval(() => {
      setCurrentDateString(Date());
    }, 1000);

    return () => {
      if (intervalId) {
        clearInterval(intervalId);
      }
    };
  }, [setCurrentDateString]);

  return (
    <span>
      <i>{text}</i>
      <span className="random-date-string">
        <b>{currentDateString}</b>
      </span>
    </span>
  );
}

const renderStaticRangeLabel = () => (
  <CustomStaticRangeLabelContent text={'This is a custom label content: '} />
);

const [state, setState] = useState([
  {
    startDate: new Date(),
    endDate: null,
    key: 'selection'
  }
]);

<DefinedRange
  onChange={item => setState([item.selection])}
  ranges={state}
  renderStaticRangeLabel={renderStaticRangeLabel}
  staticRanges={[
    {
      label: 'Hoy',
      hasCustomRendering: true,
      range: () => ({
        startDate: new Date(),
        endDate: new Date()
      }),
      isSelected() {
        return true;
      }
    }
  ]}
/>;
```
