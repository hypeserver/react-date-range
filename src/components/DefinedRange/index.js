import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from '../../styles';
import { defaultInputRanges, defaultStaticRanges } from '../../defaultRanges';
import { rangeShape } from '../DayCell';
import InputRangeField from '../InputRangeField';
import cx from 'classnames';

class DefinedRange extends Component {
  constructor(props) {
    super(props);
    this.state = {
      rangeOffset: 0,
      focusedInput: -1,
    };
  }

  handleRangeChange = range => {
    const { onChange, ranges, focusedRange } = this.props;
    const selectedRange = ranges[focusedRange[0]];
    if (!onChange || !selectedRange) return;
    onChange({
      [selectedRange.key || `range${focusedRange[0] + 1}`]: { ...selectedRange, ...range },
    });
  };

  getRangeOptionValue(option) {
    const { ranges = [], focusedRange = [] } = this.props;

    if (typeof option.getCurrentValue !== 'function') {
      return '';
    }

    const selectedRange = ranges[focusedRange[0]] || {};
    return option.getCurrentValue(selectedRange) || '';
  }

  getSelectedRange(ranges, staticRange) {
    const focusedRangeIndex = ranges.findIndex(range => {
      if (!range.startDate || !range.endDate || range.disabled) return false;
      return staticRange.isSelected(range);
    });
    const selectedRange = ranges[focusedRangeIndex];
    return { selectedRange, focusedRangeIndex };
  }

  render() {
    const {
      headerContent,
      footerContent,
      onPreviewChange,
      inputRanges,
      staticRanges,
      ranges,
      renderStaticRangeLabel,
      rangeColors,
      className,
      hideStaticInputRanges,
    } = this.props;

    return (
      <>
        {!hideStaticInputRanges && (
          <div className={cx(styles.definedRangesWrapper, className)}>
            {headerContent}
            <div className={styles.staticRanges}>
              {staticRanges.map((staticRange, i) => {
                const { selectedRange, focusedRangeIndex } = this.getSelectedRange(ranges, staticRange);
                let labelContent;

                if (staticRange.hasCustomRendering) {
                  labelContent = renderStaticRangeLabel(staticRange);
                } else {
                  labelContent = staticRange.label;
                }

                return (
                  <button
                    type="button"
                    className={cx(styles.staticRange, {
                      [styles.staticRangeSelected]: Boolean(selectedRange),
                    })}
                    style={{
                      color: selectedRange
                        ? selectedRange.color || rangeColors[focusedRangeIndex]
                        : null,
                    }}
                    key={i}
                    onClick={() => this.handleRangeChange(staticRange.range(this.props))}
                    onFocus={() => onPreviewChange && onPreviewChange(staticRange.range(this.props))}
                    onMouseOver={() =>
                      onPreviewChange && onPreviewChange(staticRange.range(this.props))
                    }
                    onMouseLeave={() => {
                      onPreviewChange && onPreviewChange();
                    }}>
                    <span tabIndex={-1} className={styles.staticRangeLabel}>
                      {labelContent}
                    </span>
                  </button>
                );
              })}
            </div>
            <div className={styles.inputRanges}>
              {inputRanges.map((rangeOption, i) => (
                <InputRangeField
                  key={i}
                  styles={styles}
                  label={rangeOption.label}
                  onFocus={() => this.setState({ focusedInput: i, rangeOffset: 0 })}
                  onBlur={() => this.setState({ rangeOffset: 0 })}
                  onChange={value => this.handleRangeChange(rangeOption.range(value, this.props))}
                  value={this.getRangeOptionValue(rangeOption)}
                />
              ))}
            </div>
            {footerContent}
          </div>
        )}
      </>
    );
  }
}

DefinedRange.propTypes = {
  inputRanges: PropTypes.array,
  staticRanges: PropTypes.array,
  hideStaticInputRanges: PropTypes.bool,
  ranges: PropTypes.arrayOf(rangeShape),
  focusedRange: PropTypes.arrayOf(PropTypes.number),
  onPreviewChange: PropTypes.func,
  onChange: PropTypes.func,
  footerContent: PropTypes.any,
  headerContent: PropTypes.any,
  rangeColors: PropTypes.arrayOf(PropTypes.string),
  className: PropTypes.string,
  renderStaticRangeLabel: PropTypes.func,
};

DefinedRange.defaultProps = {
  inputRanges: defaultInputRanges,
  staticRanges: defaultStaticRanges,
  hideStaticInputRanges: false,
  ranges: [],
  rangeColors: ['#3d91ff', '#3ecf8e', '#fed14c'],
  focusedRange: [0, 0],
};

export default DefinedRange;
