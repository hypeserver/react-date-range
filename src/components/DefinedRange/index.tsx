import React, { Component } from 'react';
import styles from '../../styles';
import { defaultInputRangesGen, defaultStaticRangesGen } from '../../defaultRanges';
import InputRangeField from '../InputRangeField';
import cx from 'classnames';
import { InputRange, InputRangeWihLabel, isSureRange, isWithRangeGen, MaybeMaybeRange, OtherRangeProps, Range, RangeFocus, StaticRange, WeekStartsOn } from '../../types';

type ComponentProps = {
  className?: string;
  focusedRange: RangeFocus;
  footerContent?: JSX.Element;
  headerContent?: JSX.Element;
  inputRanges: InputRangeWihLabel[];
  onChange?: (keyedRange: { [k: string]: Range; }) => void;
  onPreviewChange?: (r?: Range) => void;
  rangeColors: string[];
  ranges: MaybeMaybeRange[];
  renderStaticRangeLabel?: (r: StaticRange) => JSX.Element;
  staticRanges: StaticRange[];
  weekStartsOn: WeekStartsOn;
}

export type DefinedRangeProps = ComponentProps;

class DefinedRange extends Component<ComponentProps> {

  public static defaultProps = {
    ranges: [],
    inputRanges: [],
    staticRanges: [],
    rangeColors: ['#3d91ff', '#3ecf8e', '#fed14c'],
    focusedRange: [0, 0],
    weekStartsOn: 1,
  };
  // dateOptions: { weekStartsOn: any; inputRanges: InputRangeWihLabel[]; staticRanges: StaticRange[]; };

  constructor(props: ComponentProps) {
    super(props);
    // const weekStartsOn = this.props.weekStartsOn !== undefined ? this.props.weekStartsOn : 0;
    // this.dateOptions = {
    //   weekStartsOn,
    //   inputRanges:
    //     this.props.inputRanges !== undefined ? this.props.inputRanges : defaultInputRangesGen({ weekStartsOn }),
    //   staticRanges:
    //     this.props.staticRanges !== undefined
    //       ? this.props.staticRanges
    //       : defaultStaticRangesGen({ weekStartsOn }),
    // };
    this.state = {
      rangeOffset: 0,
      focusedInput: -1,
    };
  }

  handleRangeChange = (range: Range) => {
    const { onChange, ranges, focusedRange } = this.props;
    const selectedRange = ranges[focusedRange[0]];
    if (!onChange || !selectedRange) return;
    onChange({
      [selectedRange.key || `range${focusedRange[0] + 1}`]: { ...selectedRange, ...range },
    });
  }

  getRangeOptionValue(option: InputRange) {
    const { ranges = [], focusedRange = [] } = this.props;

    if (typeof option.getCurrentValue !== 'function') {
      return '';
    }

    const selectedRange = (focusedRange[0] && ranges[focusedRange[0]]) || {};
    if (!isSureRange(selectedRange)) {
      throw new Error('Bug, expecting sure range at this stage');
    }
    return option.getCurrentValue(selectedRange) || '';
  }

  getSelectedRange(ranges: Range[], staticRange: StaticRange) {
    const focusedRangeIndex = ranges.findIndex(range => {
      if (!isSureRange<OtherRangeProps>(range) || range.disabled) return false;
      return staticRange.isSelected(range);
    });
    const selectedRange = ranges[focusedRangeIndex];
    return { selectedRange, focusedRangeIndex };
  }

  render() {
    const {
      className,
      footerContent,
      headerContent,
      inputRanges,
      onPreviewChange,
      rangeColors,
      ranges,
      renderStaticRangeLabel,
      staticRanges,
    } = this.props;

    console.log('ir', inputRanges);
    console.log('sr', staticRanges);

    return (
      <div className={cx(styles.definedRangesWrapper, className)}>
        {headerContent}
        <div className={styles.staticRanges}>
          {staticRanges.map((staticRange, i) => {
            const { selectedRange, focusedRangeIndex } = this.getSelectedRange(ranges, staticRange);
            if (staticRange.hasCustomRendering && !renderStaticRangeLabel) {
              throw new Error('You should provie a renderStaticRangeLabel function when setting staticRange.hasCustomRendering = true');
            }
            if (!staticRange.label && !renderStaticRangeLabel) {
              throw new Error('Either provie a range with a label property or a renderStaticRangeLabel function to this component');
            }
            const labelContent = renderStaticRangeLabel && staticRange.hasCustomRendering
              ? renderStaticRangeLabel(staticRange)
              : staticRange.label;
            const actualStaticRange = isWithRangeGen(staticRange) ? staticRange.range(this.props) : staticRange.range as Range;

            return (
              <button
                type="button"
                className={cx(styles.staticRange, {
                  [styles.staticRangeSelected]: Boolean(selectedRange),
                })}
                style={{
                  color: selectedRange
                    ? selectedRange.color || rangeColors[focusedRangeIndex]
                    : undefined,
                }}
                key={i}
                onClick={() => this.handleRangeChange(actualStaticRange)}
                onFocus={() => onPreviewChange && onPreviewChange(actualStaticRange)}
                onMouseOver={() =>
                  onPreviewChange && onPreviewChange(actualStaticRange)
                }
                onMouseLeave={() => {
                  onPreviewChange && onPreviewChange();
                }}>
                <span tabIndex={-1} className={styles.staticRangeLabel}>
                  {labelContent}
                </span>
              </button>
            );
          })}
        </div>
        <div className={styles.inputRanges}>
          {inputRanges.map((rangeOption, i) => (
            <InputRangeField
              key={i}
              styles={styles}
              label={rangeOption.label}
              onFocus={() => this.setState({ focusedInput: i, rangeOffset: 0 })}
              onBlur={() => this.setState({ rangeOffset: 0 })}
              onChange={value => this.handleRangeChange(rangeOption.range(value, this.props))}
              value={this.getRangeOptionValue(rangeOption)}
            />
          ))}
        </div>
        {footerContent}
      </div>
    );
  }
}

export default DefinedRange;
