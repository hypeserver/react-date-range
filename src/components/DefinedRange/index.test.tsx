import React from 'react';
import { mount, shallow } from 'enzyme';

import DefinedRange from '.';
import { StaticRange } from '../../types';
import { createStaticRanges } from '../..';

describe('DefinedRange tests', () => {
  test('Should call "renderStaticRangeLabel" callback correct amount of times according to the "hasCustomRendering" option', () => {
    const renderStaticRangeLabel = jest.fn();
    const staticRanges = createStaticRanges([
      {
        label: 'Dynamic Label',
        range: {},
        hasCustomRendering: true,
      },
      {
        label: 'Static Label',
        range: {},
      },
      {
        label: 'Hede',
        range: {},
        hasCustomRendering: true,
      },
    ]);
    mount(
      <DefinedRange
        staticRanges={staticRanges}
        renderStaticRangeLabel={renderStaticRangeLabel}
      />
    );

    expect(renderStaticRangeLabel).toHaveBeenCalledTimes(2);
  });

  test('Should render dynamic static label contents correctly', () => {
    const renderItalicLabelContent = () => (
      <i className={'italic-label-content'}>{'Italic Content'}</i>
    );
    const renderBoldLabelContent = () => <b className={'bold-label-content'}>{'Bold Content'}</b>;
    const renderSomethingElse = () => <img className={'random-image'} />;

    const renderStaticRangeLabel = function(staticRange: StaticRange): JSX.Element {
      let result;

      if (staticRange.id === 'italic') {
        result = renderItalicLabelContent();
      } else if (staticRange.id === 'bold') {
        result = renderBoldLabelContent();
      } else {
        result = renderSomethingElse();
      }

      return result;
    };

    const wrapper = shallow(
      <DefinedRange
        staticRanges={createStaticRanges([
          {
            id: 'italic',
            range: {},
            hasCustomRendering: true,
          },
          {
            label: 'Static Label',
            range: {},
          },
          {
            id: 'whatever',
            range: {},
            hasCustomRendering: true,
          },
          {
            id: 'bold',
            range: {},
            hasCustomRendering: true,
          },
        ])}
        renderStaticRangeLabel={renderStaticRangeLabel}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
