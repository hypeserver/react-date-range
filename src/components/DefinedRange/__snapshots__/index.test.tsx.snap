// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefinedRange tests Should render dynamic static label contents correctly 1`] = `
<div
  className="rdrDefinedRangesWrapper"
>
  <div
    className="rdrStaticRanges"
  >
    <button
      className="rdrStaticRange"
      key="0"
      onClick={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": undefined,
        }
      }
      type="button"
    >
      <span
        className="rdrStaticRangeLabel"
        tabIndex={-1}
      >
        <i
          className="italic-label-content"
        >
          Italic Content
        </i>
      </span>
    </button>
    <button
      className="rdrStaticRange"
      key="1"
      onClick={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": undefined,
        }
      }
      type="button"
    >
      <span
        className="rdrStaticRangeLabel"
        tabIndex={-1}
      >
        Static Label
      </span>
    </button>
    <button
      className="rdrStaticRange"
      key="2"
      onClick={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": undefined,
        }
      }
      type="button"
    >
      <span
        className="rdrStaticRangeLabel"
        tabIndex={-1}
      >
        <img
          className="random-image"
        />
      </span>
    </button>
    <button
      className="rdrStaticRange"
      key="3"
      onClick={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": undefined,
        }
      }
      type="button"
    >
      <span
        className="rdrStaticRangeLabel"
        tabIndex={-1}
      >
        <b
          className="bold-label-content"
        >
          Bold Content
        </b>
      </span>
    </button>
  </div>
  <div
    className="rdrInputRanges"
  />
</div>
`;
